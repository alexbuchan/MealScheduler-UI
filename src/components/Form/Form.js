import React from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import Validator from '../../lib/Validator';

const propTypes = {
  submitButtonLabel: PropTypes.string,  // Provide a label name for the submit button
  validate: PropTypes.bool,             // Should the form be validated
  fields: PropTypes.object.isRequired,  // Provide the fields for validation
  onSubmit: PropTypes.func.isRequired,  // Provide the onSubmit function to handle form submittion,
  redirectTo: PropTypes.string,         // If redirect enabled, pass route to be redirected to
  shouldRedirect: PropTypes.func,       // If redirect enabled, pass conditions to Form for redirect
};

class Form extends React.Component {
  constructor(props) {
    super(props);

    const validationFields = this.props.children.map(child => child.props.name);
    this.validator = new Validator(...validationFields);

    this.state = {
      redirect: false,
      disableSubmitButton: false,
      validation: this.validator.valid()
    }

    this.submitted = false;
  }

  componentDidUpdate() {
    if (this.state.disableSubmitButton) {
      this.turnOffSubmitButtonDisable = setTimeout(() => {
        this.setState(() => ({ disableSubmitButton: false }))
      }, 1000);
    }

    this.redirect();
  }

  handleOnSubmit = (ev) => {
    ev.preventDefault();

    if (this.props.validate) {
      const validation = this.validator.validate(this.props.fields);
      this.setState({ validation });
      this.submitted = true;
  
      if (validation.isValid) {
        this.props.onSubmit(this.props.fields);
        this.disableSubmitButtonOnSubmit();
      }
    } else {
      this.props.onSubmit(this.props.fields);
      this.disableSubmitButtonOnSubmit();
    }
  }

  disableSubmitButtonOnSubmit = () => {
    this.setState({
      disableSubmitButton: true
    });
  }

  redirect = () => {
    if (this.props.shouldRedirect()) {
      this.setState({ redirect: true });
    }
  }

  render() {
    if (this.state.redirect) return <Redirect to={ this.props.redirectTo } />;

    let validation = this.submitted ? this.validator.validate(this.props.fields) : this.state.validation;
    const children = React.Children.toArray(this.props.children);
    const clonedChildren = children.map(child => React.cloneElement(child, { validationField: validation[child.props.name] }));

    return (
      <div className='my-form'>
        { clonedChildren }
        <button
          className="form-submit-button"
          disabled={ this.state.disableSubmitButton }
          onClick={ this.handleOnSubmit }
        >
          { this.props.submitButtonLabel }
        </button>
      </div>
    );
  }
}

Form.defaultProps = {
  submitButtonLabel: 'Submit',
  validate: false
}

Form.propTypes = propTypes;
export default Form;