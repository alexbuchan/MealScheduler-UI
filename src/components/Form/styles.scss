.my-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.form-submit-button {
  width: 178px;
  height: 40px;
  margin-top: 30px;
  background-image: linear-gradient(0deg, #4db1f3, #b0d8f3);
  border: 1px solid #4575ad;
  border-radius: 4px;
}

.form-link {
  font-size: small;
  margin-top: 10px;
}