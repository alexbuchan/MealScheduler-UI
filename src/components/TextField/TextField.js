import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  label: PropTypes.string,              // The name on the label that renders in the UI. /Optional
  name: PropTypes.string.isRequired,    // The name of the field
  type: PropTypes.string,               // The type of input for the field
  value: PropTypes.string.isRequired,   // The value of the input in the field
  placeholder: PropTypes.string,         // Optional placeholder for textfield
  onChange: PropTypes.func.isRequired,  // The onChange function for updating the field
  validationField: PropTypes.object     // Validation object returned from the Validator instance.
};

const TextField = ({ label, name, type, value, placeholder, onChange, validationField }) => {
  const inputFieldClass = () => {
    return (!validationField.isInvalid) ? "" : "input-field-error";
  }

  const withValidation = () => {
    return (
      <div className='form-element'>
        <label>{ label }</label>
        <input
          data-test='input-with-validation'
          className={ inputFieldClass() }
          name={ name }
          type={ type }
          value={ value }
          placeholder={ placeholder }
          onChange={ onChange }
        />
        <span className="help-block">{ validationField.message }</span>
      </div>
    );
  }

  const noValidation = () => {
    return (
      <div className='form-element'>
        <label>{ label }</label>
        <input
          name={ name }
          type={ type }
          value={ value }
          placeholder={ placeholder }
          onChange={ onChange }
        />
      </div>
    );
  }

  if (validationField) {
    return withValidation();
  }

  return noValidation();
}

TextField.defaultProps = {
  type: 'text'
}

TextField.propTypes = propTypes;
export default TextField;