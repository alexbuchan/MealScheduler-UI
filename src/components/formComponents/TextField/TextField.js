import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  label: PropTypes.string,              // The name on the label that renders in the UI. /Optional
  name: PropTypes.string.isRequired,    // The name of the field
  type: PropTypes.string,               // The type of input for the field
  value: PropTypes.string.isRequired,   // The value of the input in the field
  width: PropTypes.number,              // The width property for input
  height: PropTypes.number,             // The height property for input
  disabled: PropTypes.bool,             // Disables the text field
  placeholder: PropTypes.string,        // Optional placeholder for textfield
  onChange: PropTypes.func.isRequired,  // The onChange function for updating the field
  validationField: PropTypes.object,    // Validation object returned from the Validator instance.
  isRequired: PropTypes.bool,           // When validation from a form is turned on, this marks the field as required
  inline: PropTypes.bool                // Places the label inline with the textfield 
};

const TextField = ({ label, name, type, value, width, height, disabled, placeholder, onChange, validationField, inline }) => {
  const inlineLabel = () => {
    return (inline) ? "text-field-inline" : "text-field-component";
  }

  const inputFieldClass = () => {
    return (!validationField.isInvalid) ? "" : "input-field-error";
  }

  const isRequiredLabel = () => {
    if (validationField.isRequired) {
      return <span data-test='text-field-is-required' className='text-field-is-required'>*</span>;
    }

    return null;
  }

  const withValidation = () => {
    return (
      <div className={ inlineLabel() }>
        <div>
          <label>{ label }</label>
          { isRequiredLabel() }
        </div>

        <input
          data-test='input-with-validation'
          className={ `${inputFieldClass()} text-field-input` }
          name={ name }
          style={ { width, height } }
          type={ type }
          value={ value }
          disabled={ disabled }
          placeholder={ placeholder }
          onChange={ onChange }
        />

        <span className="help-block">{ validationField.message }</span>
      </div>
    );
  }

  const noValidation = () => {
    return (
      <div className={ inlineLabel() }>
        <label className='text-field-label'>{ label }</label>
        <input
          name={ name }
          className='text-field-input'
          style={ { width, height } }
          type={ type }
          value={ value }
          disabled={ disabled }
          placeholder={ placeholder }
          onChange={ onChange }
        />
      </div>
    );
  }

  if (validationField) {
    return withValidation();
  }

  return noValidation();
}

TextField.defaultProps = {
  type: 'text',
  inline: false
}

TextField.propTypes = propTypes;
export default TextField;