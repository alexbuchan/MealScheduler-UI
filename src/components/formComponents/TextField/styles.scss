.text-field-component {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.text-field-inline {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.text-field-is-required { 
  flex-direction: row;
  margin-left: 3px;
  font-size: 17px;
  color: red;
}

.text-field-label {
  margin: 0 10px 0 0;
}

.text-field-input {
  height: 41px;
  width: 200px;
  border: 1px solid lightgray;
  border-radius: 4px;
}

.text-field-input::-webkit-calendar-picker-indicator {
  opacity: 0.75;
}

.text-field-input::-webkit-datetime-edit-fields-wrapper {
  opacity: 0.3;
  margin-left: 13px;
}