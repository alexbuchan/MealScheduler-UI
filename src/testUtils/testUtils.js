import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}