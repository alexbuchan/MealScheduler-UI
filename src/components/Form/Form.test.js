import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { mount } from 'enzyme';
import { componentSetup, findByTestAttribute } from '../../testUtils/testUtils';
import Form from './Form';
import TextField from '../TextField/TextField';

describe('Form', () => {
  let instance;

  beforeEach(() => {
    const props = {
      fields: { username: 'username', password: 'password' },
      onSubmit: () => {}
    }

    instance = componentSetup(Form, props);
  });

  describe('when form has been submitted', () => {
    it('Submit button should be disabled for a second then enabled again', () => {
      jest.useFakeTimers();
      const preventDefault = jest.fn();
      const formSubmitButton = findByTestAttribute(instance, 'form-submit-button');

      formSubmitButton.simulate('click', { button: 0, preventDefault: preventDefault });
      
      jest.advanceTimersByTime(100);
      const formSubmitButton2 = findByTestAttribute(instance, 'form-submit-button');
      expect(formSubmitButton2.prop('disabled')).toBe(true);

      jest.advanceTimersByTime(999);
      const formSubmitButton3 = findByTestAttribute(instance, 'form-submit-button');
      expect(formSubmitButton3.prop('disabled')).toBe(false);
    });
  });

  describe('#render', () => {
    it('instance should NOT contain validation in state', () => {
      expect(instance.state()).not.toHaveProperty('validation');
    });

    describe('when passing only required props', () => {
      describe('when no form elements are passed to the form', () => {
        let instance;

        beforeEach(() => {
          const props = {
            fields: { username: 'username', password: 'password' },
            onSubmit: () => {}
          }

          instance = componentSetup(Form, props);
        });

        it('should render just a submit button', () => {
          expect(instance).toMatchSnapshot();
        });
      });

      describe('when a form element is passed to the form', () => {
        let instance;

        beforeEach(() => {
          const props = {
            fields: { username: 'username', password: 'password' },
            onSubmit: () => {},
            children: <TextField key={ 1 } value='' name='textField1' onChange={ () => {} } />
          }

          instance = componentSetup(Form, props);
        });

        it('should render that element and a submit button', () => {
          expect(instance).toMatchSnapshot();
        });
      });
    });

    describe('when redirect is enabled', () => {
      let instance;
        
      beforeEach(() => {
        const props = {
          fields: { username: 'username', password: 'password' },
          onSubmit: () => {},
          redirect: true,
          redirectTo: '/route',
          shouldRedirect: false
        }

        instance = mount(React.createElement(props => (
          <MemoryRouter>
            <Form {...props} />
          </MemoryRouter>
        ), props));
      });

      it('redirects to path specified in prop redirectTo', () => {
        const form = instance.find(Form);
        expect(form).toMatchSnapshot();
        instance.setProps({ shouldRedirect: true });
        const form2 = instance.find(Form);
        expect(form2).toMatchSnapshot();
      });
    });

    describe('when validation is enabled', () => {
      it('instance should contain validation in state', () => {

      });

      describe('when no form elements are passed to the form', () => {
        it('should render just a submit button', () => {

        });

        it('submit button should contain the value "Submit"', () => {

        });
      });

      describe('when a form element is passed to the form', () => {
        it('should render that element and a submit button', () => {

        });

        it('submit button should contain the value "Submit"', () => {

        });

        it('child element should contain a new prop called ValidationField', () => {

        });
      });
    });
  });

  describe('#handleOnSubmit', () => {
    describe('when validate is disabled', () => {
      it('should call props function "onSubmit"', () => {

      });

      it('should call "disableSubmitButtonOnSubmit"', () => {

      });
    });

    describe('when validate is enabled', () => {
      it('this.submitted should be true', () => {

      });

      describe('when fields are valid', () => {
        it('should call props function "onSubmit"', () => {

        });

        it('should call "disableSubmitButtonOnSubmit"', () => {

        });
      });

      describe('when fields are NOT valid', () => {
        it('should NOT call props function "onSubmit"', () => {

        });
        
        it('should NOT call "disableSubmitButtonOnSubmit"', () => {

        });
      });
    });
  });
});