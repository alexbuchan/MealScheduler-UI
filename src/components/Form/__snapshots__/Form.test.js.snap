// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form #render when passing only required props when a form element is passed to the form should render that element and a submit button 1`] = `
<div
  className="my-form"
>
  <TextField
    key="1"
    name="textField1"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    className="form-submit-button"
    data-test="form-submit-button"
    disabled={false}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`Form #render when passing only required props when no form elements are passed to the form should render just a submit button 1`] = `
<div
  className="my-form"
>
  <button
    className="form-submit-button"
    data-test="form-submit-button"
    disabled={false}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`Form #render when redirect is enabled redirects to path specified in prop redirectTo 1`] = `
<Form
  fields={
    Object {
      "password": "password",
      "username": "username",
    }
  }
  onSubmit={[Function]}
  redirect={true}
  redirectTo="/route"
  shouldRedirect={false}
  submitButtonLabel="Submit"
  validate={false}
>
  <div
    className="my-form"
  >
    <button
      className="form-submit-button"
      data-test="form-submit-button"
      disabled={false}
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
</Form>
`;

exports[`Form #render when redirect is enabled redirects to path specified in prop redirectTo 2`] = `
<Form
  fields={
    Object {
      "password": "password",
      "username": "username",
    }
  }
  onSubmit={[Function]}
  redirect={true}
  redirectTo="/route"
  shouldRedirect={true}
  submitButtonLabel="Submit"
  validate={false}
>
  <Redirect
    to="/route"
  >
    <Lifecycle
      onMount={[Function]}
      onUpdate={[Function]}
      to="/route"
    />
  </Redirect>
</Form>
`;
